### ISSUE TITLE
Port to Six

### ISSUE THREAD

**Issue by quozl** [sugarlabs/sugar-toolkit-gtk3#382]
Port the Sugar Toolkit for GTK+ 3 to Six, providing both Python 2 and Python 3 bindings from this package.

Progress goals:

- [x] fix a critical accidental dependency; remove the `from jarabe.model import shell` from `src/sugar3/graphics/popwindow.py` and use some other design,
- [x] <s>port C code in `src/sugar3`</s> _none of the C files include Python,h_,
- [x] port toolkit code in `src/sugar3` which calls C code, and test that the C code can be called from REPL,
- [x] port toolkit code in `src/sugar3/graphics`,
- [x] port graphics examples, and test that examples work,
- [x] port toolkit code in `src/sugar3/activity`,
- [x] port the [Hello World activity](https://github.com/sugarlabs/hello-world-fork),
- [x] port all remaining toolkit code, such as `src/sugar3/datastore`, `presence`, `bundle`,
- [x] port to TelepathyGLib,
- [x] port make files, and test that "make install" installs both bindings,
- [x] regenerate documentation,
- [ ] <s>make local Fedora and Debian packages for testing by others.</s>
- [X] release new version of toolkit.

_this is an issue for tracking a major project which is larger than one person's contributions, so developers should make pull requests to satisfy any of these items, and not ask to own the issue._

**Comment by yashagrawal3** [comment]
@quozl I would like to work on this issue

**Comment by quozl** [comment]
A discussion began in pull-request https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/385 about redundant work.  As your release manager, my opinion is;

* everyone who contributes to issue https://github.com/sugarlabs/sugar-toolkit-gtk3/issues/382 should cherry-pick each other's patches into a local branch before starting any work, so that you can avoid doing work that was already done.
* you should all learn how to and be comfortable enough to maintain several branches within your own local clones.
* when making a pull request, rebase your changes to master.
* a pull request may contain work by you alone, or it may contain patches by others too.
* a pull request may be for review only, or for review and merging.
* to aid in review and collaboration, you may use a branch name of _python3_.

I'm unwilling to merge any pull requests until the whole issue is resolved; as I don't want master branch in a broken state in any point of git history, because this breaks use of "git bisect".

**Comment by quozl** [comment]
Reviewed the critical dependency during a meeting.  It is probably a coding error; if PopWindow is used by an activity, the reference counted variable would not be shared with the shell, because the shell is in a different process.  The shell is the only consumer of PopWindow.  No activities use PopWindow.  So the code can be removed from the toolkit and reimplemented in src/jarabe/desktop/activitychooser.py ActivityChooser.

**Comment by Aniket21mathur** [comment]
> [ ] port the [Hello World activity](https://github.com/sugarlabs/hello-world-fork),

Ported by @pro-panda .

**Comment by Aniket21mathur** [comment]
>     * make local Fedora and Debian packages for testing by others.

Link to Debian packages (Python 2 version).
https://github.com/Aniket21mathur/sugar-toolkit-gtk3-packages

**Comment by Aniket21mathur** [comment]
>     * port toolkit code in `src/sugar3` which calls C code, and test that the C code can be called from REPL,

@pro-panda @quozl I have not found any `.so` files in the respective directory.

**Comment by Aniket21mathur** [comment]
>     * regenerate documentation,

I wanted to know which documentation is referred here, and how do we regenerate it?

**Comment by Aniket21mathur** [comment]
>     * port graphics examples, and test that examples work,

Already ported by @pro-panda. I reviewed the changes again. :-)

**Comment by pro-panda** [comment]
> > ```
> > * regenerate documentation,
> > ```
> 
> I wanted to know which documentation is referred here, and how do we regenerate it?

We generate documentation using [Sphinx](http://www.sphinx-doc.org/en/master/). It is currently hosted at https://developer.sugarlabs.org/sugar3/

You can use https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/make-doc.sh to generate documentation.
It is generated from docstrings like [this](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/activity.py#L284) and [this](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/activity.py#L20).
I briefly reviewed https://github.com/sugarlabs/sugar-toolkit-gtk3/commit/aa8a5e70c415e6c2acf4ff373d9b366ac4692bb1 and did not see any changes in the docs. I've not reviewed this in other commits yet.

**Comment by pro-panda** [comment]
> > ```
> > * port toolkit code in `src/sugar3` which calls C code, and test that the C code can be called from REPL,
> > ```
> 
> @pro-panda @quozl I have not found any `.so` files in the respective directory. Also did a manual search, no import of any `.c` file in a python file.

Thanks, python does not support importing `.c` files directly. You must compile them and generate a `.so` to use it. We had one case in [sugar-datastore](https://github.com/sugarlabs/sugar-datastore/blob/0cdf6df6c3084a6388472545493fd7ef4baae970/src/carquinyol/metadatareader.c).

I agree, there are no such cases in sugar-toolkit

**Comment by pro-panda** [comment]
> > ```
> > * port graphics examples, and test that examples work,
> > ```
> 
> Already ported by @pro-panda. I reviewed the changes again. :-)

Thanks

**Comment by Aniket21mathur** [comment]
> I briefly reviewed [aa8a5e7](https://github.com/sugarlabs/sugar-toolkit-gtk3/commit/aa8a5e70c415e6c2acf4ff373d9b366ac4692bb1) and did not see any changes in the docs.

Thanks Agreed.
I also looked for changes in https://github.com/sugarlabs/sugar-toolkit-gtk3/tree/master/doc, didn't find any, anywhere else to look?

**Comment by pro-panda** [comment]
> I also looked for changes in https://github.com/sugarlabs/sugar-toolkit-gtk3/tree/master/doc, didn't find any, anywhere else to look?

Sorry, what did you look for ?
You can go through each commit relating to the port to Python 3, which has been or will be pushed to master, and see if the changes made require changing the docs as well. If yes, please do.

Rephrasing my previous statement,
> I briefly reviewed aa8a5e7 and did not see any changes in the docs.

I briefly reviewed aa8a5e7 and did not see any instances where the docs need to be updated

**Comment by quozl** [comment]
Generated docs are for the API, therefore look for changes to API.  I don't remember any that were necessary for the port to Six, but the port to TelepathyGLib may cause some change.  I'm falling behind while travelling, so this is from memory.

**Comment by Aniket21mathur** [comment]
@quozl @pro-panda thanks for explaining. Comapred changes in https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/412/files with _sugar3.activity.activity module_ , _sugar3.presence.activity module_, _sugar3.presence.buddy module_, _sugar3.presence.connectionmanager module_, _sugar3.presence.presenceservice module_, _sugar3.presence.sugartubeconn module_, _sugar3.presence.tubeconn module_ API docs. The telepathy port does not include any changes in function names mentioned in the API docs,  nor does it contradicts their description. I see no edits to the current documentation, but their might be something that we can add, like how to use `telepathy_text_chan` in activities? Need suggestions.

**Comment by quozl** [comment]
Reviewed changes so far; we've completed documentation changes; of the changes to API that were made, the documentation did not originally describe what was changed.  Tracking the opportunity in another issue.

**Comment by sukhdeepg** [comment]
@Aniket21mathur as you quoted: 

> As the toolkit supports both Python and Python 3 we should have multi-version packages for the toolkit
I had a question:
I checked the python code in the repository, for the py files I saw, it was python3. So, how are we providing python2 support? and to completely understand the package meaning, is it what we install using `apt`(in case of Ubuntu) or the zip file which is created after `make dist`. Thanks in advance.

**Comment by pro-panda** [comment]
Hi @sukhdeepg,

this is not the best place to ask doubts. Please use the [mailing list](http://lists.sugarlabs.org/listinfo/sugar-devel).

> I checked the python code in the repository, for the py files I saw, it was python3

Which files did you see ?
A `.py` file does not enforce the Python version with which it is supposed to be executed. I'm assuming you confused it with shebang line.


> ...and to completely understand the package meaning, is it what we install using apt(in case of Ubuntu) or the zip file which is created after make dist

Yes, you usually install packages using apt, and use the `make dist` command in the process of generating commands. I've skipped a lot of details on how these work. Feel free to explore

**Comment by quozl** [comment]
Component sugar-toolkit-gtk3 branch _master_ is both Python 2 and Python 3 compatible right now.  This was achieved using Six.  Both `autogen.sh` and `configure` accept an argument to select the version of Python to build for.

Hiding subthread as resolved.  Better forum is mailing list.  But I don't think you had exhausted all sources of information before asking your question.  :grin:

**Comment by sukhdeepg** [comment]
@pro-panda @quozl my bad for not researching thoroughly. Point taken and understood 🙂

**Comment by quozl** [comment]
No worries.  I saw you subscribed to the mailing list.  Thanks!
